### PR TITLE
Use different approach in group_data.data.frame()

### LINKED ISSUE #5182: group_data.data.frame()
This is a bit esoteric, but the implementation of `group_data()` does not react well to `psych::describe`  kinds of data frames, mostly because of our use of `[, 0]` : 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(neuropsychology)
#> Loading required package: tidyverse
#> ************
#> Welcome to neuropsychology v0.5.1 (c) Dominique Makowski.
#> See documentation on https://www.rdocumentation.org/packages/neuropsychology
#> Do not hesitate to create an issue on https://github.com/neuropsychology/neuropsychology.R/issues with questions, comments, or movie recommendations.
#> ************

df <- psych::describe(select_numeric(personality))
ncol(df)
#> [1] 13
str(df)
#> Classes 'psych', 'describe' and 'data.frame':    15 obs. of  13 variables:
#>  $ vars    : int  1 2 3 4 5 6 7 8 9 10 ...
#>  $ n       : num  1302 904 1327 1292 1327 ...
#>  $ mean    : num  26.796 0.855 4.513 22.143 3.389 ...
#>  $ sd      : num  10.59 1.23 1.28 3.92 1.5 ...
#>  $ median  : num  22.9 0 5 21.3 3.5 ...
#>  $ trimmed : num  24.381 0.646 4.541 21.616 3.394 ...
#>  $ mad     : num  3.72 0 1.48 2.65 1.48 ...
#>  $ min     : num  -0.0356 0 0 15.2249 0 ...
#>  $ max     : num  80.1 5 7 56.6 7 ...
#>  $ range   : num  80.2 5 7 41.3 7 ...
#>  $ skew    : num  2.2317 1.292 -0.2954 2.4559 -0.0175 ...
#>  $ kurtosis: num  4.75 0.689 -0.657 11.983 -0.51 ...
#>  $ se      : num  0.2935 0.0408 0.0351 0.1091 0.0413 ...
str(df[, 0])
#> List of 13
#>  $ vars    : int [1:15] 1 2 3 4 5 6 7 8 9 10 ...
#>  $ n       : num [1:15] 1302 904 1327 1292 1327 ...
#>  $ mean    : num [1:15] 26.796 0.855 4.513 22.143 3.389 ...
#>  $ sd      : num [1:15] 10.59 1.23 1.28 3.92 1.5 ...
#>  $ median  : num [1:15] 22.9 0 5 21.3 3.5 ...
#>  $ trimmed : num [1:15] 24.381 0.646 4.541 21.616 3.394 ...
#>  $ mad     : num [1:15] 3.72 0 1.48 2.65 1.48 ...
#>  $ min     : num [1:15] -0.0356 0 0 15.2249 0 ...
#>  $ max     : num [1:15] 80.1 5 7 56.6 7 ...
#>  $ range   : num [1:15] 80.2 5 7 41.3 7 ...
#>  $ skew    : num [1:15] 2.2317 1.292 -0.2954 2.4559 -0.0175 ...
#>  $ kurtosis: num [1:15] 4.75 0.689 -0.657 11.983 -0.51 ...
#>  $ se      : num [1:15] 0.2935 0.0408 0.0351 0.1091 0.0413 ...
#>  - attr(*, "row.names")= chr [1:15] "Age" "Salary" "Study_Level" "BMI" ...
#>  - attr(*, "dimensions")= int 13
#>  - attr(*, "class")= chr "describe"
group_data(df)
#> Error: Input must be a vector, not a `describe` object.
```

<sup>Created on 2020-05-04 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

This breaks package `neuropsychology`

### PR DESCRIPTION
**Pull request by hadley**
Maybe fixes #5182

@romainfrancois I was coincidentally working on this!

### CODE REVIEW

**Comment by DavisVaughan** [comment]
Is there any reason we are attempting to keep the type of `.data` here? I figured the `group_data()` was an internal detail that should always return a bare tibble.

From what I can tell, `group_data.grouped_df` makes no attempt to retain the type of `.data`.

So could the implementation be this?

```r
group_data.data.frame <- function(.data) {
  rows <- as.list(seq_len(nrow(.data)))
  rows <- new_list_of(rows, ptype = integer())
  tibble::new_tibble(list(.rows = rows), nrow = 1L)
}
```

File: R/group_data.R [hunk 1]
```diff
@@ -41,7 +41,7 @@ group_data <- function(.data) {
 
 #' @export
 group_data.data.frame <- function(.data) {
-  out <- vec_init(.data[, 0], 1)
+  out <- vec_init(.data[integer()], 1)
```
Comment (romainfrancois) [inline R/group_data.R:44]: This might give another problem, where packages don't implement `[` correctly, perhaps `.data[, integer()]` ?

**Comment by romainfrancois** [comment]
It needs to make a difference between data.frame and tbl_df, this passes tests: 

```r
#' @export
group_data.data.frame <- function(.data) {
  rows <- new_list_of(list(seq_len(nrow(.data))), ptype = integer())
  if (is_tibble(.data)) {
    new_tibble(list(.rows = rows), nrow = 1L)
  } else {
    new_data_frame(list(.rows = rows), n = 1L)
  }
}
```

not sure about other packages though.

**Comment by hadley** [comment]
Oh yeah, that's definitely a better approach. I'll finish it off later today.

**Comment by hadley** [comment]
Oh you already did it 😄

**Pull request closed without merging**